### PR TITLE
[addon-resizer] feat: Automatic reload nanny configuration

### LINKED ISSUE #5700: Automatically reload nanny configuration when updated
**Which component are you using?**:

Using addon-resizer in conjunction with the metrics server 

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

Feature request is designed to improve user experience for applying nanny configuration changes.

**Describe the solution you'd like.**:

Update the addon resizer so that it will automatically reload the nanny configuration when the configuration changes.  This avoids alternative solutions noted below.

**Describe any alternative solutions you've considered.**:

We've implemented a sidecar container to monitor for nanny configuration updates.  If there is an update, the addon resizer container/nanny is restarted.  This is not ideal since the restart is not highly available.  

We've also relied on process documentation detailing that the metrics server pods must be restarted to pick up nanny configuration changes. 

**Additional context.**:

None.

### PR DESCRIPTION
**Pull request by AhmedGrati**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Introduce automatic reload of nanny configuration.

#### Which issue(s) this PR fixes:

Fixes #5700.

#### Does this PR introduce a user-facing change?

```release-note
N/A
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/autoscaler/tree/master/addon-resizer/enhancements/5700-nanny-configuration-reload

### CODE REVIEW

File: addon-resizer/nanny/main/pod_nanny.go [hunk 1]
```diff
@@ -111,10 +113,55 @@ func main() {
...
+		StopChan:       make(chan struct{}, 1),
+		ScaleDownDelay: *scaleDownDelay,
+		ScaleUpDelay:   *scaleUpDelay,
+		PollPeriod:     period,
+		HealthCheck:    hc,
+		Threshold:      uint64(*threshold),
+		Client:         k8s,
+	}
+
+	runNanny(*n, *nannyConfigurationFromFlags, *configDir)
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go:131]: nannyConfigurationFromFlags and configDir doesn't change, can we have them in the Nanny struct?

File: addon-resizer/nanny/main/pod_nanny.go [hunk 2]
```diff
@@ -111,10 +113,55 @@ func main() {
...
+			glog.Info("stopping nanny")
+			n.Stop()
+			// run nanny
+			glog.Info("reloading configuration and starting nanny")
+			runNanny(*n, *nannyConfigurationFromFlags, *configDir)
+		}
+	}
+}
+
+func runNanny(n nanny.Nanny, nannyCfdFlags nannyconfigalpha.NannyConfiguration, configDir string) {
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go:151]: Seems we could make it a nany function.

File: addon-resizer/nanny/main/pod_nanny.go [hunk 3]
```diff
@@ -111,10 +113,55 @@ func main() {
...
+	estimator, err := getNannyEstimator(n, nannyCfdFlags, configDir)
 	if err != nil {
 		glog.Fatal(err)
 	}
+	n.Estimator = estimator
+	go n.PollAPIServer()
+}
+
+func getNannyEstimator(n nanny.Nanny, nannyCfdFlags nannyconfigalpha.NannyConfiguration, configDir string) (nanny.ResourceEstimator, error) {
```
Comment (mwielgus) [inline addon-resizer/nanny/main/pod_nanny.go:160]: Seems we could make it a Nanny function.

File: addon-resizer/nanny/utils/fswatcher.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+)
+
+// FsWatcher is a wrapper helper for watching files
+type FsWatcher struct {
```
Comment (mwielgus) [inline addon-resizer/nanny/utils/fswatcher.go:29]: Can we get some tests for this, please?

**Comment by AhmedGrati** [comment]
@mwielgus Thanks for the review. I addressed your comments. For the `FsWatcher` unit tests, I added unit tests for `add` and `initWatcher` methods. For the `watch` method it needs root access to watch config files, which can be tested manually using this image `ahmedgrati/addon-resizer-amd64:1.8.19`. wdyt?

File: addon-resizer/nanny/main/pod_nanny.go (hunk 1 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/main/pod_nanny.go]: Updated!

File: addon-resizer/nanny/main/pod_nanny.go (hunk 2 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/main/pod_nanny.go]: Updated!

File: addon-resizer/nanny/utils/fswatcher.go (hunk 4 above)
Comment (AhmedGrati) [reply addon-resizer/nanny/utils/fswatcher.go:29]: Updated!

**Comment by AhmedGrati** [comment]
ping @mwielgus. Can you please drop another quick review on this?

**Pull request closed without merging**